### DISCUSSION TITLE
Importing the initial state

### DISCUSSION THREAD

**Discussion by till** [ent/ent#2951] [Q&A] [answered]
Hi 👋🏼 !

We're extending/replacing (refactoring) an existing software (Django-based) and recently started using entgo for new pieces where database is required. To be able to extend the schema with new entities and relations, we imported parts of the existing database into entgo entities like this:

```shell
go run ariga.io/entimport/cmd/entimport \
    -dsn "postgres://${DB_USERNAME}:${DB_PASSWORD}@localhost:${DB_PORT}/db?sslmode=disable" \
    -tables "table1,table2,table3"
```

So table1 through table3 are _old_, and we aded table4 and table5.

This seems to be the trick, each table is essentially an entity in the existing system.

Now, we're approaching first rollouts and are looking for ways to generate migration files of the new entities (table4 and table5) that we are added. I read about versioned migrations using Atlas, which seems to be nicer for prod than the auto-migrations, and the overall setup seems to work: as in, if I don't have anything, it'll create the migration file that we can apply.

What's unclear to me is how or where we configure it, to add initial state.

For context: we've setup the bits with `entc` and that part is working — it creates an `.sql` file with all the entities known to `entgo`.

But this also includes the entities (table1 through table3) I imported from the existing database. I tried dumping the existing structure, but then it complains about the `atlas.sum` file to be out of sync. And I can't find an import of some kind, to say, "this is the first .sql file".

Ideally, I would be able to use an export (from `pg_dump`) and say, "this is the initial state", and then I imagine 😆 being able to generate a migration on top of that.

Thanks for helping! :)

**Comment by till** [comment]
(Of course) Made some progress after I posted this:

```shell
pg_dump --schema-only ... > ./ent/migrate/migrations/TS_initial.sql
```

Then generated the `atlas.sum`:

```shell
atlas migrate --dir "file://ent/migrate/migrations" hash
```

But now, when I try to create my migration on top, I get a couple odd errors regarding syntax, etc..

```
go run -mod=mod ent/migrate/main.go foo
2022/09/23 19:12:24 failed generating migration file: postgres: schema "" was not found
exit status 1
```

This is due to `SET` in the export, but then there is another error where a `pg_dump` will `CREATE TABLE` and follow with `ALTER TABLE` and that doesn't seem to work either. Is there a better way than `pg_dump` to create the initial import?

**Comment by giautm** [comment, marked as answer]
@till , I think baseline migration is what you need: https://atlasgo.io/blog/2022/09/05/announcing-migration-execution#moving-an-existing-project-to-atlas-with-baseline-migrations

**Reply by till** [reply]
Nice, this worked much better than `pg_dump`. 👍🏼 

Now, we get different errors generating the migration. Is there a way to show more context?

```go
        err = migrate.NamedDiff(ctx, dsn, os.Args[1], opts...)
	if err != nil {
		log.Fatalf("failed generating migration file: %w", err)
	}
```

It stops with `reverse alter table "table1": unexpected attribute change (expect IDENTITY): []`. This seems similar to #2744.  Is there generally more way to inspect what it's struggling with?

**Reply by giautm** [reply]
@masseelch, Hey Jannik, can you take a look please.

**Reply by giautm** [reply]
> It stops with reverse alter table "table1": unexpected attribute change (expect IDENTITY): []. This seems similar to https://github.com/ent/ent/issues/2744. Is there generally more way to inspect what it's struggling with?

Is this error comes from Ent? when you generate the migration files? or when you inspect the current schema? Please give us more detail, including the ent schema of table1 and the current DDL in the database.

**Reply by till** [reply]
> > It stops with reverse alter table "table1": unexpected attribute change (expect IDENTITY): []. This seems similar to https://github.com/ent/ent/issues/2744. Is there generally more way to inspect what it's struggling with?
> 
> 
> 
> Is this error comes from Ent? when you generate the migration files? or when you inspect the current schema? Please give us more detail, including the ent schema of table1 and the current DDL in the database.

When I am trying to create a migration on top of the baseline. I am assuming the entities I imported are using some Postgres dialect that it doesn't understand yet.

**Reply by giautm** [reply]
Please give us the DDL of table1 and the ent schema generated by EntImport. Something may go wrong with entimport.

**Reply by till** [reply]
I can try:

So the baseline contains this:

```sql
CREATE TABLE "public"."auth_user" ("id" serial NOT NULL, "password" character varying(128) NOT NULL, "last_login" timestamptz NULL, "is_superuser" boolean NOT NULL, "username" character varying(150) NOT NULL, "first_name" character varying(30) NOT NULL, "last_name" character varying(150) NOT NULL, "email" character varying(254) NOT NULL, "is_staff" boolean NOT NULL, "is_active" boolean NOT NULL, "date_joined" timestamptz NOT NULL, PRIMARY KEY ("id"));
-- create index "auth_user_username_6821ab7c_like" to table: "auth_user"
CREATE INDEX "auth_user_username_6821ab7c_like" ON "public"."auth_user" ("username");
-- create index "auth_user_username_key" to table: "auth_user"
CREATE UNIQUE INDEX "auth_user_username_key" ON "public"."auth_user" ("username");
```

(It's part of a really old django app)

The schema is this:

```go
package schema

import (
	"time"

	"entgo.io/ent"
	"entgo.io/ent/dialect/entsql"
	"entgo.io/ent/schema"
	"entgo.io/ent/schema/edge"
	"entgo.io/ent/schema/field"
)

type AuthUser struct {
	ent.Schema
}

func (AuthUser) Fields() []ent.Field {
	return []ent.Field{
		field.Int32("id"),
		field.String("password"),
		field.Time("last_login").Optional(),
		field.Bool("is_superuser").Default(false),
		field.String("username").Unique(),
		field.String("first_name").Optional(),
		field.String("last_name").Optional(),
		field.String("email").Unique(),
		field.Bool("is_staff").Default(false),
		field.Bool("is_active").Default(true),
		field.Time("date_joined").Default(time.Time{}),
	}
}
func (AuthUser) Edges() []ent.Edge {
	return []ent.Edge{
           // edges
	}
}
func (AuthUser) Annotations() []schema.Annotation {
	return []schema.Annotation{entsql.Annotation{Table: "auth_user"}}
}
```

What else do you need?

I am happy to continue debugging, if I manage to extract more than the error above.

**Reply by masseelch** [reply]
Can you import the old tables with `atlas migrate diff` instead (give the database as target on an empty migration directory) and not with pg_dump? 

This should either give you more info or fix the the issue.

**Reply by till** [reply]
@masseelch if you read above, I didn't use `pg_dump`. I used:

```
atlas migrate diff \
--dir "file://ent/migrate/migrations" \
--to "postgres://${DB_USERNAME}:${DB_PASSWORD}@127.0.0.1:31337/db?sslmode=disable" \
--dev-url "postgres://${DB_USERNAME}:test123@127.0.0.1:5432/migration_test?sslmode=disable"
```

Generating the baseline is not the problem. It's the first migration I am trying to create on top of the baseline.

**Reply by masseelch** [reply]
> @masseelch if you read above, I didn't use `pg_dump`. I used:
> 
> ```
> atlas migrate diff \
> --dir "file://ent/migrate/migrations" \
> --to "postgres://${DB_USERNAME}:${DB_PASSWORD}@127.0.0.1:31337/db?sslmode=disable" \
> --dev-url "postgres://${DB_USERNAME}:test123@127.0.0.1:5432/migration_test?sslmode=disable"
> ```
> 
> Generating the baseline is not the problem. It's the first migration I am trying to create on top of the baseline.

Sorry, typed my answer before I saw yours.

**Reply by till** [reply]
Not to worry, thanks for looking into it.

I thinking, I need to maybe add more of the schema to the entity. As in, size, and maybe explicit types, etc.?

**Reply by masseelch** [reply]
> Not to worry, thanks for looking into it.
> 
> I thinking, I need to maybe add more of the schema to the entity. As in, size, and maybe explicit types, etc.?

I am in bed already (my time zone is late in night). I will investigate tomorrow.

**Reply by till** [reply]
Same, have a good evening!

**Reply by giautm** [reply]
In the DDL, column ID is `"id" serial NOT NULL`. But the ent schema is `field.Int32("id")`. I think this is a bug of entimport. Can you change it manual to workaround? Then generate the migration files again.

```go
func (AuthUser) Fields() []ent.Field {
	return []ent.Field{
		field.Int("id").
			SchemaType(map[string]string{
				dialect.Postgres: "serial" // Override Postgres.
			}),
```

**Reply by till** [reply]
I briefly tried it and that worked, will update my entities accordingly!

**Reply by till** [reply]
OK, I tried some more. I had to update a bunch of entities and edges - all no problem.

Thank you so much for getting me on right track! :)

I guess, last question is, is there a better way to dump errors when it tries to generate migrations? Or mostly looking closer at the initial schema and then figuring out where my entities are wrong?
